docs/design.md

```diff
@@ -0,0 +1,371 @@
+
+The choice of version and characteristics affects how the shard index is
+populated:
+
+1. CARv1: the index is calculated upon shard activation.
+2. Indexless CARv2: the index is calculated upon shard activation.
+3. Indexed CARv2: the inline index is adopted as-is as the shard index.
+
+### Shard states
+
+1. **Available:** the system knows about this shard and is capable of serving
+   data from it instantaneously, because (a) an index exists and is loaded, and
+   (b) data is locally accessible (e.g. it doesn't need to be fetched from a
+   remote mount).
+2. **Unavailable:** the system knows about this shard, but is not capable of
+   serving data from it because the shard is being initialized, or the mount is
+   not available locally, but still accessible with work (e.g. fetching from a
+   remote location).
+3. **Destroyed**: the shard is no longer available; this is permanent condition.
+
+### Operations
+
+#### Shard registration
+
+To register a new shard, call:
+
+```go
+dagst.RegisterShard(key []byte, mount Mount, opts ...RegisterOpts) error
+```
+
+1. This method takes a shard key and a `Mount`.
+2. It initializes the shard in `Unavailable` state.
+3. Calls `mount.Info()` to determine if the mount is of local or remote type.
+4. Calls `mount.Stat()` to determine if the mount target exists. If not, it
+   errors the registration.
+5. If remote, it fetches the remote resource into the scrap area.
+6. It determines the kind of CAR it is, and populates the index accordingly.
+7. Sets the state to `Available`.
+8. Returns.
+
+This method is _synchronous_. It returns only when the shard is fully indexed
+and available for serving data. This embodies the consistency property of the
+ACID model, whereby an INSERT is immediately queriable upon return.
+
+_RegisterOpts is an extension point._ In the future it can be used to pass in an
+unseal/decryption function to be used on access (e.g. such as when fast, random
+unseal is available).
+
+#### Shard destruction
+
+To destroy a shard, call:
+
+```go
+dagst.DestroyShard(key []byte) (destroyed bool, err error)
+```
+
+#### Other operations
+
+  * `[Pin/Unpin]Shard()`: retains the shard data in the local scrap area.
+  * `ReleaseShard()`: dispose of / release local scrap copies or other resources
+    on demand (e.g. unsealed copy).
+  * `[Lock/Unlock]Shard()`: for exclusive access.
+  * `[Incr/Decr]Shard()`: refcounting on shards.
+
+### Mounts
+
+Shards can be located anywhere, and can come and go dynamically e.g. Filecoin
+deals expire, removable media is attached/detached, or the IPFS user purges
+content.
+
+It is possible to mount shards with CARs accessible through the local
+filesystem, detachable mounts, NFS mounts, distributed filesystems like
+Ceph/GlusterFS, HTTP, FTP, etc.
+
+This versatility is provided by an abstraction called `mount.Mount`, a pluggable
+component which encapsulates operations/logic to:
+
+1. `Fetch() (io.ReadCloser, error)` Load a CAR from its origin.
```

Comment:
We should add a `FetchIndex` API here. If `FetchIndex` dosen't get us anything/errors out, we can build the Index based on the CAR we get from `Fetch` during Shard registration.

_Rationale:_

For an implementation like Lotus, even though Graphsyc writes out a CARv2 file, the Lotus mount `Fetch` implementation will only fetch the CARv1 payload as that's what's sealed into a sector. So,  during shard registration, what the DAG Store will see is a CARv1 payload and re-build the Index even though we had it in the original CARv2 thus doing wasteful work.

Reply:
Adding a `FetchIndex()` method here would break encapsulation as the `Mount` would now need to know how to seek to an index (which the shard management layer already knows how to process CARv1, CARv2-index, CARv2+index).

Also, all `Mount` implementations would need to duplicate this logic.

Instead, I think of the `Mount` as a low-level primitive that allows us to access remote byte streams.

I think it might be better to add a `io.Seeker` interface here that allows the shard management layer to:

1. Consume the pragma + header.
2. Decide if it wants only the index. If so, seek to the right position and read from there.